mkt/developers/views_payments.py

```diff
@@ -391,15 +391,23 @@ def in_app_products(request, addon_id, addon, webapp=True, account=None):
     products = addon.inappproduct_set.all()
     new_product = InAppProduct(webapp=addon)
     form = InAppProductForm()
-    list_url = None
-    detail_url = None
-    if addon.origin:
-        list_url = _fix_origin_link(reverse('in-app-products-list',
-                                            kwargs={'origin': addon.origin}))
-        detail_url = _fix_origin_link(reverse('in-app-products-detail',
-                                              # {guid} is replaced in JS.
-                                              kwargs={'origin': addon.origin,
-                                                      'guid': "{guid}"}))
+
+    inapp_origin = addon.origin
+    if not inapp_origin:
+        if not addon.guid:
+            raise TypeError(
+                'Cannot derive origin: no declared origin, no GUID')
+        # Derive a marketplace specific origin out of the GUID.
+        # This is for apps that do not specify a custom origin.
+        inapp_origin = 'marketplace:{}'.format(addon.guid)
```

Comment:
I'd try to avoid some of this nesting by stating the 3 cases:

```py
if addon.origin
    inapp_origin = addon.origin
elif addon.guid:
    # Derive a marketplace specific origin out of the GUID.
    # This is for apps that do not specify a custom origin.
    inapp_origin = 'marketplace:{}'.format(addon.guid)
else:
    raise TypeError('Cannot derive origin: no declared origin, no GUID')